### PR TITLE
Make name.decode stricter

### PR DESCRIPTION
**Pull request by jviide**
This pull request makes the `name.decode` function to be stricter about what kind of data it accepts:

 * Check that there's enough packet data to read before each step (reading the label header, reading label data, pointer decoding). Raise an error when there's not enough packet data.
 * Ensure that the maximum total name length is 253 (when counting the separator dots). Raise an error when it seems like the name will be longer.
 * The two most significant bits of label headers bytes must either be both set or be both unset. Raise an error when encountering a label header byte that doesn't adhere to this rule.
   * This also effectively limits the length of a single label to 63 bytes.
 * Require pointers to always point to _prior_ data (based on a reading of [RFC 1035, section 4.1.4](https://datatracker.ietf.org/doc/html/rfc1035#section-4.1.4)). Raise an error when a pointer doesn't adhere to this rule.
   * This helps avoiding infinite pointer loops (example: https://rustsec.org/advisories/RUSTSEC-2018-0007.html).
 * Fix a corner case where a valid compressed name got suffixed with two extra dots.
 * Make the whole function non-recursive.

This pull request also adds tests aimed to catch these cases.

Without to these checks it's possible to craft packets that take advantage of `name.decode` to cause Node.js to run out of memory when processing a ~60 kilobyte packet that takes advantage of infinite pointer loops and unlimited name lengths:

```js
const packet = require("dns-packet");

const buf = Buffer.alloc(65500);
buf.set([
  0xbe, 0xef, 0x85, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
]);
buf.fill(Buffer.from([0x01, 0x00]), 12);
buf.writeUInt16BE(0xc00c, buf.length - 2);

packet.decode(buf);
```

Just for full disclosure: Prior to making this pull request I confirmed from @mafintosh over Twitter DMs whether it's okay to submit this PR publicly 🙂

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -46,33 +46,53 @@ name.decode = function (buf, offset) {
...
+        // "[...] an entire domain name or a list of labels at the end of a domain name
+        // is replaced with a pointer to a prior occurance (sic) of the same name."
+        throw new Error('Can\'t decode name (bad pointer)')
+      }
+      offset = jumpOffset
+      oldOffset = jumpOffset
+      consumedBytes += jumped ? 0 : 1
+      jumped = true
+    } else {
+      throw new Error('Can\'t decode name (bad label)')
```
Comment (mafintosh) [inline index.js:90]: Cannot instead of Can't - Same for the above ones

**Comment by mafintosh** [comment]
Thanks for the fix! Do you mind fixing my nit, then it's good to go.

**Comment by jviide** [comment]
Changed "Can't" to "Cannot"

**Comment by mafintosh** [comment]
5.3.1

**Comment by pusateri** [comment]
RFC 1035 Section 2.3.1 requires labels to be 63 characters or less. You cannot change this.
